src/server.mjs

```diff
@@ -1094,17 +1118,148 @@ api.post("/getPathNames", (req, res, next) => {
+
+  // overwrites file if it already exists
+  const fileStream = fs.createWriteStream(destinationFile, { flags: 'w' });
+  await finished(Readable.fromWeb(response.body).pipe(fileStream));
+
+}
+
+const fetchAndValidate = async(url, maxBytes) => {
+  const options = {
+    method: "GET",
+    credentials: "omit",
+    cache: "default",
+    signal: timeoutController(config.fetchTimeout).signal
+  };
+
+  let response = await fetch(url, options);
+
+  // check for unsuccessful response codes
+  if (!response.ok) {
+    throw new BadRequestError("Fetch request failed: " + response.status);
+  }
+
+  // check for size specified in header
+  const contentLength = response.headers.get("Content-Length");
+
+  if (contentLength > maxBytes) {
+    throw new Error("Fetch request failed: contentLength exceeds maximum file size");
+  }
+
+  // use a reader to make sure we're not reading past the max size allowed
+  const reader = response.body.getReader();
+
+  let bytesRead = 0;
+  const dataRead = [];
+
+  while (true) {
+    let {done, value} = await reader.read();
+
+    if (done) {
+      break;
+    }
+
+    dataRead.push(value);
+    bytesRead += new Blob([value]).size;
+
+    if (bytesRead > maxBytes) {
+      reader.cancel();
+      throw new Error("Fetched request failed, maximum file size exceeded: " + maxSize);
+    }
+  }
+
+  return new Response(new Blob(dataRead), { headers: response.headers });
+
+}
+
+// downloads files fo the specified chunk name from the destination of the bedURL
+// includeContent only downloads the tracks.json file when set to false
+const retrieveChunk = async(baseURL, chunk, includeContent) => {
+
+  // path to the designated chunk in the temp directory
+  const chunkDir = path.resolve(config.tempDirPath, sanitize(chunk));
+
+  // TODO: check if this chunk has been downloaded before, to prevent duplicate fetches
+  if (!fs.existsSync(chunkDir)) {
+    fs.mkdirSync(chunkDir, { recursive: true });
+  } 
+
+  // baseURL/dirName/"chunk_contents.txt"
+  let chunkContentURL = new URL(path.join(sanitize(chunk), "chunk_contents.txt"), baseURL).toString();
+
+  let response = await fetchAndValidate(chunkContentURL, config.maxFileSizeBytes);
+
+  const chunkContent = await response.text();
+  const fileNames = chunkContent.split("\n");
+
+  // download all the files in the chunk
+  for (const fileName of fileNames) {
+
+    // baseURL/dirName/fileName
+    let chunkFileURL = new URL(path.join(sanitize(chunk), sanitize(fileName)), baseURL).toString();
+    
+    // download only the tracks.json file if the inlcudeContent flag is false
+    if (includeContent || fileName == "tracks.json") {
+      await downloadFile(chunkFileURL, chunkDir, fileName);
+    }
+  }
+
+  // write the baseURL to a file
+  const baseURLFile = path.resolve(chunkDir, "baseURL.txt");
+  fs.writeFile(baseURLFile, baseURL, function (error) {
+    if (error){
+      throw error;
+    }
+  })
```

Comment:
I don't understand why this system is necessary.

When the user asks to get chunk data for a tube map rendering, they send along the BED URL, and we are already finding the right region in the BED file to get the chunk information so that we know where to look on the server for the downloaded chunk data. So we can always make the base URL of the chunk with a `new URL()` based on the BED file's own URL and whatever it lists as the path to the chunk, right? Why do we need to make it once and save it in a flat file for every chunk?